server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs

```diff
@@ -346,31 +352,53 @@ onStart env serverEnv wsConn (StartMsg opId q) = catchAndIgnore $ do
   execPlanE <- runExceptT $ E.getResolvedExecPlan env logger pgExecCtx
                {- planCache -} userInfo sqlGenCtx sc scVer queryType httpMgr reqHdrs (q, reqParsed)
 
-  (telemCacheHit, execPlan) <- either (withComplete . preExecErr requestId) return execPlanE
+  (_telemCacheHit, execPlan) <- either (withComplete . preExecErr requestId) return execPlanE
   let execCtx = E.ExecutionCtx logger sqlGenCtx pgExecCtx {- planCache -} sc scVer httpMgr enableAL
 
   case execPlan of
-    E.QueryExecutionPlan queryPlan asts ->
-      case queryPlan of
-        E.ExecStepDB (tx, genSql) -> Tracing.trace "Query" $
-          execQueryOrMut timerTot Telem.Query telemCacheHit (Just genSql) requestId $
-            fmap snd $ Tracing.interpTraceT id $ executeQuery reqParsed asts (Just genSql) pgExecCtx tx
-        E.ExecStepRemote (rsi, opDef, _varValsM) ->
-          runRemoteGQ timerTot telemCacheHit execCtx requestId userInfo reqHdrs opDef rsi
-        E.ExecStepRaw (name, json) ->
-          execQueryOrMut timerTot Telem.Query telemCacheHit Nothing requestId $
-          return $ encJFromJValue $ J.Object $ Map.singleton (G.unName name) json
-    E.MutationExecutionPlan mutationPlan ->
-      case mutationPlan of
-        E.ExecStepDB (tx, _) -> Tracing.trace "Mutate" do
+    E.QueryExecutionPlan queryPlan asts -> Tracing.trace "Query" $ do
+      let cacheKey = QueryCacheKey reqParsed $ _uiRole userInfo
+      -- We ignore the response headers (containing TTL information) because
+      -- WebSockets don't support them.
+      (_responseHeaders, cachedValue) <- Tracing.interpTraceT id $ cacheLookup asts cacheKey
+      case cachedValue of
+        Just cachedResponseData -> do
+          sendSuccResp cachedResponseData $ LQ.LiveQueryMetadata 0
+        Nothing -> do
+          conclusion <- runExceptT $ forWithKey queryPlan $ \fieldName -> \case
+            E.ExecStepDB (tx, genSql) -> doQErr $ Tracing.trace "Postgres Query" $ do
+              logQueryLog logger q (Map.singleton (G.unsafeMkName fieldName) . Just <$> genSql) requestId
+              (telemTimeIO_DT, (resp)) <- Tracing.interpTraceT id $ withElapsedTime $
+                hoist (runQueryTx pgExecCtx) tx
+              return $ ResultsFragment telemTimeIO_DT Telem.Local resp []
+            E.ExecStepRemote (rsi, opDef, varValsM) -> do
+              runRemoteGQ fieldName execCtx requestId userInfo reqHdrs opDef rsi varValsM
+            E.ExecStepRaw json ->
+              buildRaw json
+          buildResult Telem.Query timerTot requestId conclusion
+          case conclusion of
+            Left _ -> pure ()
+            Right results ->
+              Tracing.interpTraceT id $ cacheStore cacheKey (encJFromInsOrdHashMap (fmap rfResponse results))
+      sendCompleted (Just requestId)
+
```

Comment:
This chunk of code seems *incredibly* similar to the corresponding piece of logic in the HTTP module. Could we get some sharing here, or is the duplication incidental?

Reply:
No, you're absolutely right. For what it's worth, there should have already been code sharing before this PR. And I did try to deduplicate some code here. But I got stuck juggling monad stacks. So I didn't manage to do this in a reasonable time.

(Note that there are some mild operational differences: e.g. while the WebSocket code writes the results to the socket, the HTTP code returns it as a Haskell value.)

Reply:
How difficult do you think it would be to do that bit of refactoring as part of this PR? If we’re going to do it at some point, now seems as good a time as any. Of course, if it’s really hard, that’s one thing, but it’s not clear to me why it would be.

Reply:
We're thinking of doing it while working on data-loader, since it will mean changing this logic significantly.